R/score_model_out.R

```diff
@@ -113,129 +119,56 @@ score_model_out <- function(model_out_tbl, target_observations, metrics = NULL,
-  metrics <- switch(output_type,
-    quantile = scoringutils::metrics_quantile(),
-    pmf = scoringutils::metrics_nominal(),
-    mean = scoringutils::metrics_point(select = "se_point"),
-    median = scoringutils::metrics_point(select = "ae_point"),
-    NULL # default
-  )
-  if (is.null(metrics)) {
-    # we have already validated `output_type`, so this case should not be
-    # triggered; this case is just double checking in case we add something new
-    # later, to ensure we update this function.
-    supported_types <- c("mean", "median", "pmf", "quantile") # nolint object_use_linter
-    cli::cli_abort(
-      "Provided `model_out_tbl` contains `output_type` {.val {output_type}};
-      hubEvals currently only supports the following types:
-      {.val {supported_types}}"
-    )
-  }
+get_metrics <- function(forecast, output_type, select = NULL) {
+  forecast_type <- class(forecast)[1]
 
-  return(metrics)
-}
-
-
-#' Convert character vector of metrics to list of functions
-#'
-#' @inheritParams score_model_out
-#'
-#' @return a list of metric functions as required by scoringutils::score()
-#'
-#' @noRd
-get_metrics_character <- function(metrics, output_type) {
-  if (output_type == "quantile") {
+  # process quantile metrics separately to allow better selection of interval
+  # coverage metrics
+  if (forecast_type == "forecast_quantile") {
     # split into metrics for interval coverage and others
-    interval_metric_inds <- grepl(pattern = "^interval_coverage_[[:digit:]][[:digit:]]$", metrics)
-    interval_metrics <- metrics[interval_metric_inds]
-    other_metrics <- metrics[!interval_metric_inds]
+    interval_metric_inds <- grepl(pattern = "^interval_coverage_", select)
+    interval_metrics <- select[interval_metric_inds]
+    other_metrics <- select[!interval_metric_inds]
 
-    # validate metrics
-    valid_metrics <- c("ae_median", "wis", "overprediction", "underprediction", "dispersion")
-    invalid_metrics <- other_metrics[!other_metrics %in% valid_metrics]
-    error_if_invalid_metrics(
-      valid_metrics = c(valid_metrics, "interval_coverage_XY"),
-      invalid_metrics = invalid_metrics,
-      output_type = output_type,
-      comment = c("i" = "NOTE: `XY` denotes the coverage level, e.g. {.val interval_coverage_95}.")
-    )
+    other_metric_fns <- scoringutils::get_metrics(forecast, select = other_metrics)
 
-    # assemble metric functions
+    # assemble interval coverage functions
     interval_metric_fns <- lapply(
       interval_metrics,
       function(metric) {
-        level <- as.integer(substr(metric, 19, 20))
+        level_str <- substr(metric, 19, nchar(metric))
+        level <- suppressWarnings(as.numeric(level_str))
+        if (is.na(level) || level <= 0 || level >= 100) {
+          stop(paste(
+            "Invalid interval coverage level:", level_str,
+            "- must be a number between 0 and 100 (exclusive)"
+          ))
```

Comment:
In this house, we can afford a little syntactic sugar with our errors.

```suggestion
          cli::cli_abort(c(
            "Invalid interval coverage level: {level_str}",
            "i" = "must be a number between 0 and 100 (exclusive)"
          ))
```